Week1 dice throw

DanaRavestein opened the pull request.

DanaRavestein committed:
start with throwing of dice + added image of scorecard

DanaRavestein committed:
Roll dice and create scorecard

DanaRavestein committed:
Merge branch 'master' into week1_dice_throw

DanaRavestein committed:
Solved issues with categories and it checks the scorecard

DanaRavestein committed:
Shows the scorecard to player + edit to assigning points

Made sure the player can see the scorecard and it updates during the game. Added the bonus points and made sure that if no category is possible the player can 'cross out' the category.

merope95 commented:
I see that you're using many classes. The code seems to work just fine but it looks a bit complicated to me. It seems that the majority of the classes exist to calculate the score for every dice combination. I think it will be much simpler if you wrote all that in a single function which calculates the score for a given dice combination. Here's an example:

```python
def get_score(dice_throw):
    if ...:  # a condition for aces
        return 3
    elif ...:  # a condition for twos
        return 6
    ...

    return 0  # if none of the above apply
```

I think this has a few benefits over your current structure:

1. The code is simpler and easier to follow and maintain
2. The code is easier to unit test
3. I think the code will be much shorter too

Also, if you want to return more stuff than just the score, you can return a dictionary instead. For example:

```python
{
	"score": 25,
	"name": "Full House",
}
```

merope95 commented:
Also, I noticed that many of the checks (for aces, tows, threes etc.) are very similar, with a difference of a single number. I suggest to extract that into a common function and only pass that number as an argument. That way you'll avoid code duplication and the code will be shorter and neater  🦉

merope95 commented:
Overall, looks very good! I made some suggestions but it looks your solution is correct as it is and the code just needs a bit of tudying up. Looking forward to seeing the whole game in action ^_^

carolinstr commented on yahtzee.py:
```diff
@@ -0,0 +1,28 @@
+import pygame as pg
+
+'''  Check classes tomorrow for DiceSet?
+class Dice:
```
Cool that you are making use of classes!

carolinstr commented on yahtzee.py:
```diff
@@ -0,0 +1,28 @@
+    if i == 0:
+        throw_again_idx = range(0, len(dice))
+    else:
+        throw_again_idx = np.array([int(i) - 1 for i in input().split()])
```
Maybe add a comment here to explain what you're doing. How I understand it you assign a different number of throws depending on if it's the player's first turn or not?

DanaRavestein replied:
Yeah, I should definitely make more comments haha. Forgot to add them in my enthusiasm when the code was finally working

carolinstr commented on yahtzee.py:
```diff
@@ -0,0 +1,28 @@
+    for j in throw_again_idx:
+            dice[j] = randint(1, 6)   
+            
+    print('Do you want to keep these dice, press k?: ', dice)  
```
Maybe:
print('These are your dice:', dice)
input('Do you want to keep them? If no, press "k"')

DanaRavestein replied:
Yes, thanks! The questions are not yet really thought through, because I am opting to use something like clicks on the dices on a table or whatever in the final game :)

carolinstr commented on yahtzee.py:
```diff
@@ -1,7 +1,10 @@
+
 
  
  #define the Dice class
```
I think it's very impressive you're using classes. But for each function in the class I would add a small comment to explain what e.g. __init__ / roll / etc does. For you it's probably super intuitive but for a reader (or at least for me) it would help if you explain it

carolinstr commented on yahtzee.py:
```diff
@@ -197,28 +246,44 @@ def __init__(self):
             
         print('You rolled: ', player1.current_throw)
```
This loop is very clever and it's nice that you implemented the option to throw again

carolinstr commented on yahtzee.py:
```diff
@@ -197,28 +246,44 @@ def __init__(self):
+ 
     for i, category in enumerate(scorecard.categories):
         category.is_possible(player1.current_throw)
         category.score(player1.current_throw)
```
Here I see why you defined all the classes separately and I think this is an awesome feature. This way you can easily access the is_possible and score, etc

carolinstr reviewed (commented):
Your code seems really thought-through and complete. I think it's very clever to define all the different types of scores as classes beforehand because you can easily access them later. I would like it if you could add some comments about the different functions and methods/ attributes so that there is no room for confusion. For example .is_possible vs .possiblescores.

carolinstr commented on yahtzee.py:
```diff
@@ -0,0 +1,289 @@
+        self.bonus = 0
+        self.total = 0
+        for i in range(0, len(self.categories)):
+            self.categoryname.append(scorecard.categories[i].name)
```
I think this should be:
self.categoryname.append(self.categories[i].name)

carolinstr commented:
After reading @merope95 comment I think maybe it makes sense to look into if it's possible to shorten the code. Maybe it is possible to write a class and under def __init__(self) make a for loop for all possible outcomes and so that self.name differs? But then you would also have to adjust the is.possible and score and I don't know if that is in the end more confusing than having different classes. For now I think the current code is very clear

DanaRavestein committed:
Processed feedback + made multiplayer possible

The classes do not contain duplicated code anymore, and are in a different file. So, now you can import them. Made a start with using pygame. Also, I made changes to what the classes do (keep scorecard per player, score function which calculates a score per category, printing the scorecard per player per roll).
